### ISSUE TITLE
Bug: typescript-eslint v7 config file fails eslint validation

### ISSUE THREAD

**Issue by NotArpit** [typescript-eslint/typescript-eslint#8478]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-eslint

### Repro Code

```TypeScript
/* eslint.config.js */
// @ts-check

import eslint from '@eslint/js';
import tseslint from 'typescript-eslint';

export default tseslint.config(
  eslint.configs.recommended,
  ...tseslint.configs.recommended,
);

/* .eslintignore */
**/types.ts
```

### ESLint Config

```javascript
// @ts-check

import eslint from '@eslint/js';
import tseslint from 'typescript-eslint';

export default tseslint.config(
  eslint.configs.recommended,
  ...tseslint.configs.recommended,
);
```

### tsconfig

```jsonc
Ran `tsc --init` in a new project.
```

### Expected Result

The `import tseslint from 'typescript-eslint'` line has no errors from the language-server

### Actual Result

The `import tseslint from 'typescript-eslint'` line errors from the language-server about `Cannot find module 'typescript-eslint' or its corresponding type declarations.ts(2307)`. Removing the `// @ts-check` directive fixes this error however I don't believe this is the expected behaviour.

### Additional Info

I can consistently reproduce this in a new project with the `package.json` deps below. Happy to provide additional details :)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`                | `^7.0.1` |
| `typescript`                           | `^5.3.3` |
| `eslint`                                   | `^8.56.0` |
| `node`                                    | `20.11.0` |
| `pnpm`                                   | `8.15.2` |
